// @ngxs/store/esm5/src/configs/messages.config.js
var _a;
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var VALIDATION_CODE = {
    STATE_NAME: 'STATE_NAME',
    STATE_UNIQUE: 'STATE_UNIQUE',
    STATE_NAME_PROPERTY: 'STATE_NAME_PROPERTY',
    STATE_DECORATOR: 'STATE_DECORATOR',
    INCORRECT_PRODUCTION: 'INCORRECT_PRODUCTION',
    INCORRECT_DEVELOPMENT: 'INCORRECT_DEVELOPMENT',
    SELECT_FACTORY_NOT_CONNECTED: 'SELECT_FACTORY_NOT_CONNECTED',
    ACTION_DECORATOR: 'ACTION_DECORATOR',
    SELECTOR_DECORATOR: 'SELECTOR_DECORATOR',
};
export { VALIDATION_CODE };
/** @type {?} */
export var CONFIG_MESSAGES = (_a = {},
    _a[VALIDATION_CODE.STATE_NAME] = (/**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return name + " is not a valid state name. It needs to be a valid object property name.";
    }),
    _a[VALIDATION_CODE.STATE_NAME_PROPERTY] = (/**
     * @return {?}
     */
    function () { return "States must register a 'name' property"; }),
    _a[VALIDATION_CODE.STATE_UNIQUE] = (/**
     * @param {?} current
     * @param {?} newName
     * @param {?} oldName
     * @return {?}
     */
    function (current, newName, oldName) {
        return "State name '" + current + "' from " + newName + " already exists in " + oldName;
    }),
    _a[VALIDATION_CODE.STATE_DECORATOR] = (/**
     * @return {?}
     */
    function () { return 'States must be decorated with @State() decorator'; }),
    _a[VALIDATION_CODE.INCORRECT_PRODUCTION] = (/**
     * @return {?}
     */
    function () {
        return 'Angular is running in production mode but NGXS is still running in the development mode!\n' +
            'Please set developmentMode to false on the NgxsModule options when in production mode.\n' +
            'NgxsModule.forRoot(states, { developmentMode: !environment.production })';
    }),
    _a[VALIDATION_CODE.INCORRECT_DEVELOPMENT] = (/**
     * @return {?}
     */
    function () {
        return 'RECOMMENDATION: Set developmentMode to true on the NgxsModule when Angular is running in development mode.\n' +
            'NgxsModule.forRoot(states, { developmentMode: !environment.production })';
    }),
    _a[VALIDATION_CODE.SELECT_FACTORY_NOT_CONNECTED] = (/**
     * @return {?}
     */
    function () {
        return 'SelectFactory not connected to store!';
    }),
    _a[VALIDATION_CODE.ACTION_DECORATOR] = (/**
     * @return {?}
     */
    function () {
        return '@Action() decorator cannot be used with static methods';
    }),
    _a[VALIDATION_CODE.SELECTOR_DECORATOR] = (/**
     * @return {?}
     */
    function () { return 'Selectors only work on methods'; }),
    _a);
